library/std/src/io/cursor.rs

```diff
@@ -396,38 +396,74 @@ fn slice_write_vectored(
     Ok(nwritten)
 }
 
-// Resizing write implementation
-fn vec_write<A>(pos_mut: &mut u64, vec: &mut Vec<u8, A>, buf: &[u8]) -> io::Result<usize>
-where
-    A: Allocator,
-{
-    let pos: usize = (*pos_mut).try_into().map_err(|_| {
+/// Reserves the required space, and pads the vec with 0s if necessary.
+/// Truncates if pos is less than vec.len()
+fn reserve_and_pad<A: Allocator>(
+    pos_mut: &mut u64,
+    vec: &mut Vec<u8, A>,
+    buf_len: usize,
+) -> io::Result<usize> {
+    let pos = (*pos_mut).try_into().map_err(|_| {
         io::const_io_error!(
             ErrorKind::InvalidInput,
             "cursor position exceeds maximum possible vector length",
         )
     })?;
-    // Make sure the internal buffer is as least as big as where we
-    // currently are
-    let len = vec.len();
-    if len < pos {
-        // use `resize` so that the zero filling is as efficient as possible
-        vec.resize(pos, 0);
-    }
-    // Figure out what bytes will be used to overwrite what's currently
-    // there (left), and what will be appended on the end (right)
-    {
-        let space = vec.len() - pos;
-        let (left, right) = buf.split_at(cmp::min(space, buf.len()));
-        vec[pos..pos + left.len()].copy_from_slice(left);
-        vec.extend_from_slice(right);
+
+    if vec.len() < pos + buf_len {
+        // allocate data upfront
+        vec.reserve(pos + buf_len - vec.len());
+    }
+    // written like this to avoid an extra alloc when using `resize()`
+    if vec.len() < pos {
+        // Safety: we have allocated enough capacity for this.
+        // And we are only writing, not reading
+        unsafe {
+            // compiles to memset
+            for i in 0..pos - vec.len() {
+                vec.spare_capacity_mut().get_unchecked_mut(i).write(0);
+            }
+            vec.set_len(pos);
+        }
     }
 
+    Ok(pos)
+}
+
+/// Writes the slice to the vec without allocating
+/// # Safety: vec must have buf.len() spare capacity
+unsafe fn vec_write_unchecked<A>(pos: usize, vec: &mut Vec<u8, A>, buf: &[u8]) -> usize
+where
+    A: Allocator,
+{
+    vec.as_mut_ptr().add(pos).copy_from(buf.as_ptr(), buf.len());
+    pos + buf.len()
+}
+
+// Resizing write implementation
+fn vec_write<A>(pos_mut: &mut u64, vec: &mut Vec<u8, A>, buf: &[u8]) -> io::Result<usize>
```

Comment:
I can definitely do that. I just copied the existing comment 😅

Padding confused me too. Cursor is slightly different from just a vec write since it has its own position. This means its possible to write to an initialised vec at position 0, and its possible to write to an empty vec at position 40